### DISCUSSION TITLE
Use a defined single nameserver?

### DISCUSSION THREAD

**Discussion by jompa-github** [MichaCo/DnsClient.NET#148] [Ideas]
Hello, fistly thinks for what looks like a really useful tool! I have tried but failed to define a single name server (by IP-address). The background is that I would like to compare results from multiple known name servers. I therefore need to point at just one server at the time. IDK if this is a feature already or something that could be added?
The records I will be querying are not A or PTR and as such, I have yet to find a client that can do this async...

**Comment by MichaConrad** [comment]
What have you tried so far? 
You can either initialize a LookupClient with one server or use Query overloads which accept a server... so lots of options to do what you need I guess?

**Reply by jompa-github** [reply]
Thanks for quick reply!

I cannot figure out how to do this as I am only getting "'NameServers' is a ReadOnly property." I am probably doing it wrong...
I am using PowerShell BTW :)

$LookupClient = New-Object -TypeName DnsClient.LookupClient
$LookupClient.NameServers = ([System.Net.IPAddress]"194.168.4.100", [int]53)
Exception:  'NameServers' is a ReadOnly property.
 
$DnsQueryAndServerOptions = New-Object -TypeName DnsClient.DnsQueryAndServerOptions
$DnsQueryAndServerOptions.NameServers = ([System.Net.IPAddress]"194.168.4.100", [int]53)
Exception:  'NameServers' is a ReadOnly property.

**Reply by MichaConrad** [reply]
I'm not familiar enough with Powershell to test this out and give you a full example so you have to figure that out yourself.

As you noticed by that error message, the list of servers cannot be changed after the client has been instantiated.
But you can pass in a configuration object of type [`LookupClientOptions`](https://github.com/MichaCo/DnsClient.NET/blob/dev/src/DnsClient/DnsQueryOptions.cs#L331) to the constructor of LookupClient.
The New-Object thing allows to pass in an `-ArgumentList`. 

So, maybe try create a New-Object of type LookupClientOptions and then pass that into the arguments list of the New-Object call to create the LookupClient.

**Reply by jompa-github** [reply]
Hello again, thanks again for your prompt replies!

I tried this;
$NameServer = [DnsClient.NameServer]::new([System.Net.IPAddress]"194.168.4.100", [int]53)

$LookupClientOptions = New-Object -TypeName DnsClient.LookupClientOptions -Property @{
    NameServers = $NameServer
    }

I am getting a "The value supplied is not valid, or the property is read-only" exception. Given the $NameServer object is OK, I guessing that "NameServers" is a R/O attribute again? There is always the possibility that I am messing this up due to ignorance or incompetence - again...

I am used to dig and nslookup where I am able to point to different DNS servers. None of these tools does what your code does - async... Brilliant and really quick! If I only can figure out how to point it to a specific name server.

//John

**Reply by jompa-github** [reply]
Hi Michael,

I apologise if I am pestering you which is not my intention. I would normally not consider myself inept or incapable of figuring stuff out but I am drawing a blank here. I am unable to define or even add a custom name server because all the NameServer object I see are read-only. I guess I am missing something here but fail to see what. The name server is usually one of the first attributes you want to see or change in a DNS client but I cannot figure this one out.

Any help or pointers would be greatly appreciated!

//John

**Reply by MichaConrad** [reply]
Maybe try writing what you want to do in a good language first, like c#. Maybe then it's easier to understand what's going on.
With IDE help the library is really easy to use. I guess you just struggle because PowerShell kinda sucks 😉

Anyways, I think you tried to create a new DnsLookupOptions object but then tried to change the nameserver property afterwards.
The api only allows you to change it via constructor Parameter. So, just pass it in at creation time..
